src/axolotl/utils/callbacks.py

```diff
@@ -317,3 +320,250 @@ def on_evaluate(
+                        # prompt_token_ids = batch['input_ids'][i][label_non_zero_indices]
+                        # prompt_token_ids = batch['input_ids'][i]
+
+                        prompt_token_ids = batch['input_ids'][i][prompt_token_indices]
+                        completion_token_ids = batch['input_ids'][i][completion_token_indices]
+
+                        # prompt_texts = tokenizer.batch_decode(batch.data['input_ids'])
+                        # prompt_texts = tokenizer.batch_decode(prompt_token_ids)
+                        prompt_text = tokenizer.decode(prompt_token_ids)
+                        completion_text = tokenizer.decode(completion_token_ids)
+
+                        completion_logit = logit[completion_token_indices]
+                        # predicted_tokens = logits_to_tokens(logit)
+                        predicted_tokens = logits_to_tokens(completion_logit)
+
+                        # Append the predicted tokens to the preds list
+                        # pred_tokens.extend(predicted_tokens)
+                        # pred_string = " ".join(predicted_tokens) # FIXME: missing spaces
+                        prediction_text = tokenizer.decode(predicted_tokens)
+
+                        # generate new prediction with trainer.model which is a transformer model
+                        # Generate new prediction with trainer.model which is a transformer model
+                        with torch.no_grad():
+                            # new_prediction = trainer.model(batch['input_ids'][i].unsqueeze(0))
+                            # new_prediction = trainer.model(prompt_token_ids.unsqueeze(0))
+                            # new_prediction = trainer.model(prompt_token_ids.unsqueeze(0))
+
+                            generation_config = GenerationConfig(
+                                repetition_penalty=1.1,
+                                # max_new_tokens=1024,
+                                # max_new_tokens=256,
+                                max_new_tokens=128,
+                                temperature=0.9,
+                                # top_p=0.95,
+                                # top_k=40,
+                                bos_token_id=tokenizer.bos_token_id,
+                                eos_token_id=tokenizer.eos_token_id,
+                                pad_token_id=tokenizer.pad_token_id,
+                                # do_sample=True,
+                                do_sample=False,
+                                use_cache=True,
+                                return_dict_in_generate=True,
+                                output_attentions=False,
+                                output_hidden_states=False,
+                                output_scores=False,
+                            )
+                            # streamer = TextStreamer(tokenizer)
+                            new_prediction = trainer.model.generate(
```

Comment:
**Help Wanted**

Not sure this is what should be used to generate the prediction either.

:x: I noticed `trainer.prediction_step` appeared strange, rarely human readable. This may be because it is used for SFT, not human readable evaluation. In which case I thought we could have both.

✅ Appears to get a reasonable output.
:x: Very slow. It appears slower than `prediction_step` or `trainer.model(batch['input_ids'][i].unsqueeze(0))` I was using previously/above.

Reply:
I think with the prediction step, what happens is the unlabeled positions usually emit junk since it's not training on those, but once you get to the response part, it's better
As for the speed, I'm not really well-versed on that part

Reply:
For speed, I was able to batch up multiple generate requests and that helped a lot! Much better now.